File: Java/Java_Core/ListaExerciciosJavaPOO/Ex003/src/main/java/com/mycompany/ex003/BombaCombustivel.java
package com.mycompany.ex003;

public class BombaCombustivel {

    private String tipoCombustivel;
    private double valorLitro;
    private double qtdCombustivel;

    public BombaCombustivel(String tipoCombustivel, double valorLitro, double qtdCombustivel) {
        this.tipoCombustivel = tipoCombustivel;
        this.valorLitro = valorLitro;
        this.qtdCombustivel = qtdCombustivel;
    }

    public void abastecerPorValor(double valorAbastecimento) {
        double quantidade = 0;
        quantidade = valorAbastecimento / this.valorLitro;
        if (this.qtdCombustivel >= quantidade) {
            System.out.println("Litros abastecidos: " + quantidade + " L.");
            this.qtdCombustivel -= quantidade;
            System.out.println("Combustivel restante: " + this.qtdCombustivel);
        } else {
            System.out.println("Sem combutível suficiente.");
        }
    }

    public void abastecerPorLitro(double litrosAbastecimento) {
        if (litrosAbastecimento > this.qtdCombustivel) {
            System.out.println("Sem combutível suficiente.");
        } else {
            System.out.println("Valor a pagar: R$"
                    + (litrosAbastecimento * valorLitro) + " por " + litrosAbastecimento
                    + " litros.");
            this.qtdCombustivel -= litrosAbastecimento;
            System.out.println("Combustivel restante: " + this.qtdCombustivel);
        }
    }

    public void alterarValor(double novoValor) {
        this.valorLitro = novoValor;
        System.out.println("Valor do(a) " + tipoCombustivel + " alterado para R$"
                + this.valorLitro);
    }

    public void alterarTipoCombustivel(String novoTipoCombustivel) {
        this.tipoCombustivel = novoTipoCombustivel;
        System.out.println("Tipo do combustível alterado para " + this.tipoCombustivel);
    }

    public void alterarQtdCombustivel(double novoEstoque) {
        if (novoEstoque >= 0) {
            this.qtdCombustivel = novoEstoque;
            System.out.println("Estoque de " + tipoCombustivel + " alterado para "
                    + this.qtdCombustivel);
        } else {
            System.out.println("Quantidade negativa não permitida.");
        }
    }

}

File: Python/Alura/pythonOO/Ex03/bomba_combustivel.py
from posto_combustivel import PostoCombustivel
from tipo_combustivel import TipoCombustivel


class BombaCombustivel(PostoCombustivel):
    def __init__(self, combustivel: TipoCombustivel, valor_litro, quantidade_estoque):
        self._combustivel = combustivel
        super().__init__(valor_litro, quantidade_estoque)

    def alterar_tipo_combustivel(self, novo_combustivel: TipoCombustivel):
        self._combustivel = novo_combustivel
        print(f"Combustível altado para {novo_combustivel.name}")

    def abastecer_por_litro(self, litros):
        if litros > super().quantidade:
            print("Estoque insuficiente.")
        else:
            print(f"Abasteceu {litros}l e pagou R$ {round(litros * super().valor, 2)}")

    def __str__(self):
        return f"Combustivel: {self._combustivel.name} - Valor do Litro: R${round(super().valor, 2)} - Estoque: {super().quantidade}l"
